R/util-MakeSnapshotHelpers.R

```diff
@@ -0,0 +1,203 @@
+#' Compile the results summary of all kri's in a snapshot
+#'
+#' @param lResults `list` the output of `Study_Assess()` containing results of kri analysis
+#'
+#' @import dplyr
+#'
+#' @export
+#'
+#' @keywords internal
+CompileResultsSummary <- function(lResults){
+  output <- lResults %>%
+    purrr::map_df(., function(kri) {
+      bind_rows(kri$lResults$lData$dfSummary)
+    }, .id = "kri") %>%
+    filter(!is.na(Flag)) %>%
+    mutate(flag_color = case_when(Flag %in% c(2, -2) ~ "red",
+                                  Flag %in% c(1, -1) ~ "amber",
+                                  Flag == 0 ~ "green"))
+
+  return(output)
+}
+
+
+#' Extract flag counts by site or KRI
+#'
+#' @param lResults `list` the output of `Study_Assess()` containing results of kri analysis
+#' @param group `character` a character field to specify what to use to group flags by. options = "site", "kri"
+#'
+#' @import dplyr
+#' @importFrom tidyr pivot_wider
+#'
+#' @export
+#'
+#' @keywords internal
+ExtractFlags <- function(lResults, group){
+  data <- CompileResultsSummary(lResults)
+  if(!group %in% c("site", "kri")){
+    stop("`group` argument must be either 'site' or 'kri'")
+  }
+  if(group == "site"){
+    results <- filter(data, str_detect(kri, "kri"))
+    if(length(results) == 0) {
+      cli::cli_alert_warning("lResults argument in `ExtractFlags()` didn't contain any KRI's with site level results, unable to group by 'site'
+                             Returning results for `kri` groups instead")
+      results <- data
+    }
+  } else {
+    results <- data
+  }
+  output <- results %>%
+    {if(group == "site") {group_by(., GroupID, flag_color)} else .} %>%
+    {if(group == "kri") {group_by(., kri, flag_color)} else .} %>%
+    summarise(n_flags = n(), .groups = "drop") %>%
+    filter(flag_color %in% c("red", "amber")) %>%
+    tidyr::pivot_wider(names_from = "flag_color", values_from = "n_flags") %>%
+    {if(!"amber" %in% names(.)) tibble::add_column(., "amber" = NA) else .} %>%
+    {if(!"red" %in% names(.)) tibble::add_column(., "red" = NA) else .} %>%
+    {if(group == "site") {select(.,"siteid" = "GroupID",
+                                 "num_of_at_risk_kris" = "amber",
+                                 "num_of_flagged_kris" = "red")}
+      else .} %>%
+    {if(group == "kri") {select(., "kri_id" = "kri",
+                                "num_of_sites_at_risk" = "amber",
+                                "num_of_sites_flagged" = "red")}
+      else .}
```

Comment:
I reviewed your PRs out of order, but think that we should refactor this part slightly to make it more concise/easily read.